src/cirrus/lib2/events.py

```diff
@@ -0,0 +1,88 @@
+import json
+import os
+from logging import Logger, getLogger
+from typing import Any, Dict, Tuple
+
+import boto3
+
+from .statedb import StateDB
+from .utils import SNSPublisher
+
+logger = getLogger(__name__)
+
+
+class WorkflowEventManager:
+    """A class for managing payload state change events, inclusing storage of
+    state(dynamo), activity(timestream), and notifications of decisions and/or status
+    changes (SNS)."""
+
+    wf_event_topic_arn = os.getenv("CIRRUS_WF_EVENT_TOPIC_ARN", None)
+
+    _message_template: Tuple[Tuple[str, Any]] = (
+        ("payload_id", "some_id"),
+        ("workflow_name", "some_workflow"),
+        ("event_type", None),
+        ("payload", None),
+    )
+
+    @classmethod
+    def _get_message(cls, event_type: str, payload: Dict) -> Dict:
+        message = dict(cls._message_template)
+        message["event_type"] = event_type
+        message["payload"] = payload
+        return json.dumps(message)
+
+    def __init__(self, logger: Logger = None, boto3_session: boto3.Session = None):
+        self.boto3_session = boto3_session if boto3_session else boto3.Session()
+        self.wfet_publisher = (
+            SNSPublisher(
+                self.wf_event_topic_arn, logger=logger, boto3_session=self.boto3_session
+            )
+            if self.wf_event_topic_arn
+            else None
+        )
+        self.statedb = StateDB(session=self.boto3_session)
+
+    def announce(self, event_type: str, payload: Dict, extra_message: str = "") -> None:
+        # WorkflowEvent Topic, if configured
+        if self.wfet_publisher:
+            message = self._get_message(event_type, payload)
+            self.wfet_publisher.add(message)
+
+    def claim_processing(self, payload):
+        self.statedb.claim_processing(payload["id"])
+        self.announce("CLAIMED_PROCESSING", payload)
+
+    def started_processing(self, payload: Dict, execution_arn: str):
+        self.statedb.set_processing(self["id"], execution_arn)
+        self.announce("STARTED_PROCESSING", payload, execution_arn)
+
+    def already_processing(self, payload: Dict):
+        self.logger.warning(f"already in processing state: {payload['id']}")
+        self.announce("ALREADY_PROCESSING", payload)
+
+    def already_completed(self, payload: Dict):
+        # TODO: check with jarrett on how he sees this function used.  perhaps
+        #      raising a special exception from `StateDB.set_completed`, which forwards
+        #      over to this function?
```

Comment:
I think you'd call this from the `else` of the `if`/`elif`/`else` at the end of `ProcessPayloads.process()`. That said, the `else` as currently written handles multiple extant states, not just `COMPLETED`, so would likely need to be broken out into separate `elif` blocks for each possible state we can handle there. We should also use the state enum instead of strings for the state matching, I think.

Reply:
I think the states we'd handle here are at least `COMPLETED`, `INVALID`, and `PROCESSING`, but its worth considering if others would be possible.

Reply:
Yeah, seems like skipping is handled the same for all payload_id states.  So, I propose this move to a `skipping` method which takes `payload`, and `state: StateEnum`, and calls `announces(f"ALREADY_{state}", payload)`.  That leaves things open to calling `skipping` on payloads in any state, which seems reasonable.